### PR TITLE
KAFKA-5126: Implement KIP-98 transactional methods in the MockProducer

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
Call for review @apurvam @hachikuji @guozhangwang @dguy

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 1]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
-    public void initTransactions() {
+    private <T> ExtendedSerializer<T> ensureExtended(Serializer<T> serializer) {
+        return serializer instanceof ExtendedSerializer ? (ExtendedSerializer<T>) serializer : new ExtendedSerializer.Wrapper<>(serializer);
+    }
 
+    @Override
+    public void initTransactions() {
+        verifyProducerState();
+        if (this.transactionInitialized) {
+            throw new IllegalStateException("MockedProducer got already initialized for transactions.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: nit: "MockProducer has already been initialized"

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 2]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
 
+    @Override
     public void sendOffsetsToTransaction(Map<TopicPartition, OffsetAndMetadata> offsets,
-                                  String consumerGroupId) throws ProducerFencedException {
-
+                                         String consumerGroupId) throws ProducerFencedException {
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
+            throw new IllegalStateException("There is not open transaction to be committed.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: "There is no..." ?

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 3]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
+        }
+        uncommittedOffsets.putAll(offsets);
     }
 
+    @Override
     public void commitTransaction() throws ProducerFencedException {
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
+            throw new IllegalStateException("There is not open transaction to be committed.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: As above

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 4]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
+        this.transactionInFlight = false;
 
     }
 
+    @Override
     public void abortTransaction() throws ProducerFencedException {
-        
+        verifyProducerState();
+        verifyTransactionsInitialized();
+        if (!transactionInFlight) {
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: ok - same thing three times. Maybe extract it to a method `verifyTransactionInflight`

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 5]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
+        this.transactionAborted = true;
+        this.transactionInFlight = false;
     }
-        
-    private <T> ExtendedSerializer<T> ensureExtended(Serializer<T> serializer) {
-        return serializer instanceof ExtendedSerializer ? (ExtendedSerializer<T>) serializer : new ExtendedSerializer.Wrapper<>(serializer);
+
+    private void verifyProducerState() {
+        if (this.closed) {
+            throw new IllegalStateException("MockedProducer is already closed.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: MockedProducer -> MockProducer

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 6]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
+        }
+        if (this.producerFenced) {
+            throw new ProducerFencedException("MockProducer got fenced.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: "MockProducer is fenced"

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 7]
```diff
@@ -117,29 +128,94 @@ public MockProducer() {
...
+        }
+    }
+
+    private void verifyTransactionsInitialized() {
+        if (!this.transactionInitialized) {
+            throw new IllegalStateException("MockedProducer did got initialized for transactions.");
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: "Mock Producer hasn't been initialized for transactions" ?

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java [hunk 8]
```diff
@@ -212,29 +295,68 @@ public void close() {
 
     @Override
     public void close(long timeout, TimeUnit timeUnit) {
-        if (closed) {
+        if (this.closed) {
```
Comment (dguy) [inline clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: Do we need to add `this`? Also below

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 9]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+        producer.initTransactions();
+        assertTrue(producer.transactionInitialized());
+    }
+
+    @Test
+    public void shouldThrowOnInitTransactionIfProducerAlreadyInitializedForTransactions() {
+        producer.initTransactions();
+        try {
+            producer.initTransactions();
+            fail("Should have trows as producer got initialized already");
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: "Should have thrown as producer is already initiailzed"

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 10]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+    @Test(expected = IllegalStateException.class)
+    public void shouldThrowOnSendOffsetsToTransactionIfTransactionsNotInitialized() {
+        producer.sendOffsetsToTransaction(null, null);
+    }
+
+    @Test
+    public void shouldThrowOnSendOffsetsToTransactionTransactionIfNoTransactionGotStarted() {
+        producer.initTransactions();
+        try {
+            producer.sendOffsetsToTransaction(null, null);
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:145]: `fail(...)`

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 11]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+    @Test(expected = IllegalStateException.class)
+    public void shouldThrowOnCommitIfTransactionsNotInitialized() {
+        producer.commitTransaction();
+    }
+
+    @Test
+    public void shouldThrowOnCommitTransactionIfNoTransactionGotStarted() {
+        producer.initTransactions();
+        try {
+            producer.commitTransaction();
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:159]: `fail(..)`

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 12]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+    @Test(expected = IllegalStateException.class)
+    public void shouldThrowOnAbortIfTransactionsNotInitialized() {
+        producer.abortTransaction();
+    }
+
+    @Test
+    public void shouldThrowOnAbortTransactionIfNoTransactionGotStarted() {
+        producer.initTransactions();
+        try {
+            producer.abortTransaction();
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java:183]: `fail(..)`

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 13]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+        producer.abortTransaction();
+
+        Map<String, Map<TopicPartition, OffsetAndMetadata>> expectedResult = new HashMap<>();
+        expectedResult.put(group, groupCommit);
+
+        assertThat(producer.consumerGroupOffsetsHistory(), equalTo(Collections.singletonList(expectedResult)));
+    }
+
+    @Test
+    public void shouldThrowOnInitTransactionIfProducerGotClosedAlready() {
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: Here and everywhere else. Got doesn't sound right. Maybe something like:
`shouldThrowOnInitTransactionIfProducerIsClosed`

File: clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java [hunk 14]
```diff
@@ -98,6 +106,425 @@ public void testManualCompletion() throws Exception {
...
+        producer.close();
+        try {
+            producer.initTransactions();
+            fail("Should have thrown as producer got closed already");
```
Comment (dguy) [inline clients/src/test/java/org/apache/kafka/clients/producer/MockProducerTest.java]: "... as producer is closed"

Same elsewhere

File: clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java (hunk 8 above)
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/clients/producer/MockProducer.java]: It's the overall code style in client code base -- I just adopt to it.

**Comment by mjsax** [comment]
Updated this.

**Review by dguy — APPROVED** [review]
LGTM

**Review by guozhangwang — APPROVED** [review]
Merged to trunk.

**Pull request closed without merging**